agimus_controller/tests/test_mpc_unicycle.py

```diff
@@ -0,0 +1,337 @@
+            )
+        )
+
+    @property
+    def ocp_results(self):
+        return self._results
+
+    @property
+    def debug_data(self):
+        return self._debug_data
+
+
+class WarmStartUnicycle(WarmStartBase):
+    def __init__(self):
+        super().__init__()
+
+    def generate(self, initial_state, reference_trajectory):
+        x0 = initial_state.robot_configuration
+        if self._previous_solution is None or len(self._previous_solution.states) == 0:
+            # No previous solution. Use the reference trajectory.
+            x_init = [wpoint.robot_configuration for wpoint in reference_trajectory[1:]]
+            us = None
+        else:
+            # Use the previous solution as warm start. For the last point, use the reference trajectory.
+            x_init = list(self._previous_solution.states[2:]) + [
+                reference_trajectory[-1].robot_configuration
+            ]
+            us = self._previous_solution.feed_forward_terms
+        return x0, x_init, us
+
+    def setup(self):
+        pass
+
+
+def unicycle_integrate(q, dq, dt):
+    x, y, theta = q
+    v, w = dq
+    return q + np.array([v * np.cos(theta), v * np.sin(theta), w]) * dt
+
+
+def unicycle_plot(x):
+    import matplotlib.pyplot as plt
+
+    sc, delta = 0.1, 0.1
+    a, b, th = x[0], x[1], x[2]
+    c, s = np.cos(th), np.sin(th)
+    refs = [
+        plt.arrow(
+            a - sc / 2 * c - delta * s,
+            b - sc / 2 * s + delta * c,
+            c * sc,
+            s * sc,
+            head_width=0.05,
+        ),
+        plt.arrow(
+            a - sc / 2 * c + delta * s,
+            b - sc / 2 * s - delta * c,
+            c * sc,
+            s * sc,
+            head_width=0.05,
+        ),
+    ]
+    return refs
+
+
+def unicycle_plot_solution(xs, figIndex=1, show=True):
+    import matplotlib.pyplot as plt
+
+    plt.figure(figIndex, figsize=(6.4, 6.4))
+    for x in xs:
+        unicycle_plot(x)
+    plt.axis([-2, 2.0, -2.0, 2.0])
+    if show:
+        plt.show()
+
+
+class TestMPCUnicycle(unittest.TestCase):
+    def test_hoziron_and_dt(self):
+        ocp = OCPUnicycle(100, 0.01)
+
+        assert ocp.horizon_size == 100
+        assert ocp.dt == 0.01
+
+    def test_reference_trajectory(self):
+        ocp = OCPUnicycle(100, 0.01)
+
+        warm_start = WarmStartUnicycle()
+        mpc = MPC()
+        mpc.setup(
+            ocp,
+            warm_start,
+            TrajectoryBuffer(
+                [
+                    (
+                        1,
+                        100,
+                    ),
+                ]
+            ),
+        )
+
+        dt_ns = int(ocp.dt * 1e9)
+
+        N_iter = 500
+
+        # TODO For some reasons, the current implementation of MPC.run first uses `clear_past`
+        # before solving the OCP.solve so the first element in the loop is removed.
+        # If the behavior is changed, the k-1 below can be replaced by k.
```

Comment:
@MaximilienNaveau Not sure if it is desired that the first element of the trajectory buffer is discarded before use.